java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java

```diff
@@ -320,9 +344,11 @@ private enum State {
   }
 
   private static class Data {
+
     // State only moves forward, never backward (WAITING_FOR_SOURCES->COLLECTING->TOO_MANY_CHANGES)
     State state = State.WAITING_FOR_SOURCES;
     Set<File> changedSources = new HashSet<>();
+    Set<File> changedNonCompilableSources = new HashSet<>();
```

Comment:
Can we have here a boolean? Should we?
Con- It will be less symmetric and elegant, shouldn't save a lot of memory since we don't expect many different proto files being edited at the same time per target
Pro- having here a Set<File> caused me to think maybe something will be done with it

Reply:
Initially, I wanted to do a boolean. However, it's again the case of state being WAITING_FOR_SOURCES, but having some files changed already. If you take a look at setSources method, once the status changes to COLLECTING, the files from this set get checked agains the sources. So if there was a change in proto file, which was not from the sources, if the status is WAITING_FOR_SOURCES, the full recompile would happen (just in case, as we cannot know, if they were from sources), but after sources were set, it would be removed in updateChangedSources method.  

So if we had a boolean, in case of changed file not from sources that happened, before the sources were collected, a full recompile would still happened, which is less correct. It's probably a rare situation in real life, but I chose to do it correctly, I don't know if you agree that it's worth it. 

The memory saving is not really relevant even if a lot of proto files would be edited, as if you check the line 375, the files stop getting checked if any proto source change was already detected

Reply:
Right... 
Now I see. 
I think we should document it.
That is indeed a good reason.
We need to accumulate since before we know about the sources we just accumulate and wait. And afterwards we intersect.
👍